releasenotes/notes/custom-serializers-qpy-0097ab79f239fcfc.yaml

```diff
@@ -0,0 +1,78 @@
+---
+features:
+  - |
+    Added a new kwarg, ``metadata_serializer`` to the
+    :func:`.qpy_serialization.dump` function for specifying a custom
+    ``JSONEncoder`` subclass for use when serializing the
+    :attr`.QuantumCircuit.metadata` attribute and a dual kwarg
+    ``metadata_deserializer`` to the :func:`.qpy_serialization.load` function
+    for specifying a ``JSONDecoder`` subclass. By default the
+    :func:`~qiskit.circuit.qpy_serialization.dump` and
+    :func:`~qiskit.circuit.qpy_serialization.load` functions will attempt to
+    JSON serialize and deserialize with the stdlib default json encoder and
+    decoder. Since :attr`.QuantumCircuit.metadata` can contain any Python
+    dictionary, even those with contents not JSON serializable by the default
+    encoder, will lead to circuits that can't be serialized. The new
+    ``metadata_serializer`` argument for
+    :func:`~qiskit.circuit.qpy_serialization.dump` enables users to specify a
+    custom ``JSONEncoder`` that will be used with the internal ``json.dump()``
+    call for  serializing the :attr`.QuantumCircuit.metadata` dictionary. This
+    can then be paired with the new ``metadata_deserializer`` argument of the
+    :func:`.qpy_serialization.load` function to decode those custom JSON
+    encodings. If ``metadata_serializer`` is specified on
+    :func:`~qiskit.circuit.qpy_serialization.dump` but ``metadata_deserializer``
+    is not specified on :func:`~qiskit.circuit.qpy_serialization.load` calls
+    the QPY will be loaded, but the circuit metadata may not be reconstructed
+    fully.
+
+    For example if you wanted to define a custom serialization for metadata and
+    then load it you can do something like::
+
+        from qiskit.qpy import dump, load
+        from qiskit.circuit import QuantumCircuit, Parameter
+        import json
+
+        class CustomObject:
+            """Custom string container object."""
+
+            def __init__(self, string):
+                self.string = string
+
+            def __eq__(self, other):
+                return self.string == other.string
+
+        class CustomSerializer(json.JSONEncoder):
+            """Custom json encoder to handle CustomObject."""
+
+            def default(self, o):  # pylint: disable=invalid-name
+                if isinstance(o, CustomObject):
+                    return {"__type__": "Custom", "value": o.string}
+                return json.JSONEncoder.default(self, o)
+
+        class CustomDeserializer(json.JSONDecoder):
+            """Custom json decoder to handle CustomObject."""
+
+            def object_hook(self, o):  # pylint: disable=invalid-name,method-hidden
```

Comment:
Do we need the pylint markers in the release note?  I didn't think we'd be checking those blocks.  If not, possibly best to get rid of them to make things cleaner for readers.

Reply:
Ah good catch, no we don't. I just copied this from the unittests verbatim Let's remove them